bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/connection/Clip2Bridge.java

```diff
@@ -0,0 +1,980 @@
+        this.bridgeHandler = bridgeHandler;
+        this.hostName = hostName;
+        this.applicationKey = applicationKey;
+        baseUrl = String.format(FORMAT_URL_RESOURCE, hostName);
+        eventUrl = String.format(FORMAT_URL_EVENTS, hostName);
+        registrationUrl = String.format(FORMAT_URL_REGISTER, hostName);
+    }
+
+    /**
+     * Send a ping to the Hue bridge to check that the session is still alive.
+     */
+    private void checkAlive() {
+        if (onlineState == State.CLOSED) {
+            return;
+        }
+        logger.debug("checkAlive()");
+        Session session = http2Session;
+        if (Objects.nonNull(session)) {
+            session.ping(new PingFrame(false), Callback.NOOP);
+        }
+        if (Instant.now().isAfter(sessionExpireTime)) {
+            fatalError(this, Http2Error.TIMEOUT);
+        }
+    }
+
+    /**
+     * Connection is ok, so reschedule the session check alive expire time. Called in response to incoming ping frames
+     * from the bridge.
+     */
+    protected void checkAliveOk() {
+        logger.debug("checkAliveOk()");
+        sessionExpireTime = Instant.now().plusSeconds(CHECK_ALIVE_SECONDS * 2);
+    }
+
+    /**
+     * Close the connection.
+     */
+    @Override
+    public void close() {
+        synchronized (this) {
+            logger.debug("close()");
+            boolean notifyHandler = (onlineState == State.ACTIVE) && !restarting;
+            onlineState = State.CLOSED;
+            closeCheckAliveTask();
+            closeSession();
+            try {
+                http2Client.stop();
+            } catch (Exception e) {
+                // ignore
+            }
+            if (notifyHandler) {
+                bridgeHandler.onConnectionOffline();
+            }
+        }
+    }
+
+    /**
+     * Close the check alive task if necessary.
+     */
+    private void closeCheckAliveTask() {
+        logger.debug("closeCheckAliveTask()");
+        ScheduledFuture<?> task = checkAliveTask;
+        if (Objects.nonNull(task)) {
+            task.cancel(true);
+        }
+        checkAliveTask = null;
+    }
+
+    /**
+     * Close the HTTP 2 session if necessary.
+     */
+    private void closeSession() {
+        logger.debug("closeSession()");
+        Session session = http2Session;
+        if (Objects.nonNull(session)) {
+            session.close(0, null, Callback.NOOP);
+        }
+        http2Session = null;
+    }
+
+    /**
+     * Method that is called back in case of fatal stream or session events. Note: under normal operation, the Hue
+     * Bridge sends a 'soft' GO_AWAY command every nine or ten hours, so we handle such soft errors by attempting to
+     * silently close and re-open the connection without notifying the handler of an actual 'hard' error.
+     *
+     * @param cause the entity that caused this method to be called.
+     * @param error the type of error.
+     */
+    private void fatalError(Object cause, Http2Error error) {
```

Comment:
This needs to be synchronized, since `fatalErrorDelayed()` will schedule each error in a separate thread and `FAILURE` will inevitably cause `CLOSE`. It's also a bit unfortunate that onFailure() reasons aren't logged at all.

My suggestion for this would be something like https://github.com/maniac103/openhab2-addons/commit/56c1e4f5fb94586c99b6e56553eae9dd5d713bcc

Reply:
Ok.